### ISSUE TITLE
Render failed on high resolution

### ISSUE THREAD

**Issue by MightyLight** [mfeemster/fractorium#71]
Today I tried to render at 4000x2000 for the first time, on 1000 quality (using openCL) and it failed just before finishing (took almost 2 min). It threw this error:

```
EmberCLns::RendererCL<float,float>::ClearBuffer

EmberCLns::RendererCL<float,float>::RunLogScaleFilter
```

When I retried it failed inmediatly and threw this:

```
EmberCLns::RendererCL<float,float>::ClearBuffer

EmberCLns::RendererCL<float,float>::ClearHist

EmberCLns::RendererCL<float,float>::RunIter

Zero iterations ran, rendering failed, aborting.
```

After lowering the resolution, it threw the first errors again, just before finishing, and again the second one on retry.

Sorry if this was already mentioned in another issue, I failed to see it.

**Comment by mfeemster** [comment]
Please provide the specs for your video card.

**Comment by MightyLight** [comment]
I have an XFX ATI Radeon R9 270X, it has 2GB GDDR5,1050 Boost Core Clock (MHz), 1280 Stream Processors, 256 Bit Memory, and 5600 Memory Clock (MHz).
I'm running Fractorium 0.9.9.4 on Windows 8.1.

Closing Fractorium let's me render again, but if I render something big on 4 supersample it happens again and doesn't let me render anymore.

**Comment by mfeemster** [comment]
Information with questions below, please answer them to help me assist you.

You are likely running out of memory. The final render dialog should give you a message in the text box telling you this is happening whenever you adjust the dimensions or supersample fields.

`1)Are you seeing this?`

You can use the strips feature to solve this problem. It splits the render into pieces. Keep in mind this will take longer to finish.

For reference, the memory used is:

(((w * ss) * (h * ss) * 16) * 2) + (w * h * 4)

^^Note the usage of ss, it scales the memory required by (4 * ss).

Plus some additional space for the gutter around the edges (which the user never sees).

So at 4000 * 2000, you are using:

256,000,000 + 32,000,000 = 288,000,000 + gutter.

This easily fits within your available memory, however if you add supersample, you will start to run out:

ss=2: (256M * 4) + 32M = 1056M. You should still be good here
ss=3: (256M * 9) + 32M = 2336M. You are out of memory at this point.

Further, you don't get the entire card's memory space made available to you. Some of that space is reserved by the driver to do display-related work.

`2) So are you seeing it fail even with an supersample of 1? Or does it only happen once you start to increase that number?`

As for recovery, OpenCL is unforgiving about recovering. It sometimes requires a restart of the program, or switch back to CPU rendering, then back to GPU to reinitialize itself.

I am currently working on a website that will have all of this documented.

**Comment by MightyLight** [comment]
1) No, the text box doesn't warn me at all, the textbox is blank while adjusting any parameter.
2) It works with ss = 1, so you are right about the problem being about memory and OpenCL not recovering. The first one was my guess as well but I didn't know it couldn't recover from there and thought it was a Fractorium problem.
 Thanks a lot for your attention.

**Comment by mfeemster** [comment]
Hmm, the text box should always tell you when your planned render will exceed available memory, I'll look into it.

In the meantime, can you try using a strips value > 1 and see if your large render completes? You can use a low quality value, like 10, to save time while testing.

Thanks.

**Comment by MightyLight** [comment]
Okey, this is quite interesting. Tried to render a 4000x4000 render with ss = 4 , strips = 5 and quality = 10 (it displays a memory usage of around 1.700.000.000) I also had Doble precision enabled.
ETA was around 20 sec at start but decreased quickly. Faster than the bar was completing. at ETA = 0,  the rendering failed, with the Iteration bar at 20%. 
This is what the textbox displays:
```
Rendering failed.

EmberCLns::RendererCL<double,float>::ClearBuffer

EmberCLns::RendererCL<double,float>::RunDensityFilterPrivate

EmberCLns::RendererCL<double,float>::RunDensityFilter

All renders completed in: 10.26s.
```

**Comment by mfeemster** [comment]
Try different combinations of ss and strips until you get a value that works and let me know what you find.

Btw, ss of 4 is overkill and is unlikely to yield anything more visually appealing than 2 or 3. TBH, 3 rarely provides any value either. 2 is what I usually go with unless there is a valid reason to go higher.

**Comment by MightyLight** [comment]
Okey, 4000x4000 with ss = 2 works with strips = 1. With ss = 3 on the other hand failed with any value of strips < 10.

On another note, on this high resolutions, the whole program lags a lot, I don't know if that is expected, I would obviusly use it on lower resolution and upgrade it when preparing the final render.

Finally, I decided to test the limiting resolution for ss = 2 and strips = 1. Found it quite easily with 8000x8000.


```
Rendering failed.


EmberCLns::RendererCL<float,float>::Alloc

Histogram, accumulator and samples buffer allocations failed, aborting.

```

Of course this exceeds my memory at 8.000.000.000. 

With ss = 1 this works, and with ss = 2 and strips < 4 this fails again with this message:

```
Rendering failed.

EmberCLns::RendererCL<float,float>::Alloc

Histogram, accumulator and samples buffer allocations failed, aborting.


EmberCLns::RendererCL<float,float>::ClearBuffer

EmberCLns::RendererCL<float,float>::ClearHist

EmberCLns::RendererCL<float,float>::RunIter

Zero iterations ran, rendering failed, aborting.

```
 Strips = 4 renders successfully. (also = 3 of course)

**Comment by mfeemster** [comment]
Sounds like you've mostly figured it out.

> on this high resolutions, the whole program lags a lot

Are you doing these high res renders in the main window, or in the final render dialog? Either way, a massive amount of processing is being done, so system slowdown is expected.

As a general rule, get your image how you want in the main window at a reasonable resolution. Then use the final render dialog to do a high quality render. There is scaling functionality there and also a preview to always show you what your final image will look like.

> With ss = 3 on the other hand failed with any value of strips < 10.

The strips value must divide evenly into the height of the buffers. If it does not evenly divide, there will be a number to the right of strips in parentheses which shows the next closest number that will be used.

**Comment by MightyLight** [comment]
I was rendering on the final render dialog, the lag was when using the main window. 
I never got the hang of the scaling functionality but I will give it a try. 

I'm aware of the strips need of divide evenly the buffers height, with strips < 10 I meant that it fails with strips = 2, 4 and 8, and strips = 10 works. 

I guess I'm closing this. Thanks for your attention Matt.

**Comment by mfeemster** [comment]
Sounds good. Sorry about the lack of documentation, it'll all be included on my website in a few months.

**Comment by mfeemster** [comment]
I am reopening this because:

When you said earlier that you were not seeing any warning text that your render would exceed available memory, alarms went off in my head since I had that working over a year ago.

I checked, and sure enough, that functionality has been broken since I did the multi-gpu work last summer. Totally slipped by me for no valid reason other than... I forgot to do it.

I've implemented a fix and will close this when I commit. Huge thanks for catching this bug.
